protocol.md

```diff
@@ -870,6 +902,12 @@ export interface TextDocumentClientCapabilities {
 ```typescript
 interface ClientCapabilities {
 	/**
+	 * The client supports receiving the result solely through $/partialResult notifications for requests from the client to the server.
+	 * If true, the server should send all data through $/partialResult and send null as the result in the final response.
```

Comment:
> I don't think the "this makes the receiver code less complicated" part is true. Clients will need to support servers that don't support streaming capabilities, so they will need to write this code anyway.

Just like the server's protocol layer, the client's protocol layer can support servers that don't support streaming by transforming the full responses to a single replace op, so the method handlers would always get a stream and don't have to differentiate. But if you do decide it on a per-method basis, the client would have to differentiate per-method and then either transform it like I just described or pass the responsibility of differentiation to the method handlers. Either way, I don't think we gain anything from doing config per-method, but it would certainly bloat the configuration and message handling.
Another small point to keep in mind: If we do it per-method, we should put it in the `RegistrationOptions` for each method, which is only in v3. All our LS and proxy code is written for LSP v2.

My idea is that eventually all responses will be streamed (it's just a matter of time for implementing it) and we should optimize for that case.

Regarding your proposal:
What do we gain from putting another layer on top of the jsonrpc layer, with own IDs, an own layer of method dispatching? The notifications certainly must still be ordered, because that is the nature of JSON Patches.
Imo it feels weird to send a notification like this
```ts
{
  method: '$/streamRequest',
  params: {
    id: 1,
    method: 'textDocument/definition',
    params: {
      uri: ...
    }
  }
}
```
with duplicated `params` and `method`. It kind of reminds me of SOAP, wrapping the actual request data in some kind of "envelope" - takes the simplicity out of JSON / JSON RPC.